lines.js

```diff
@@ -35,34 +51,60 @@ function buttonState(e) {
         button1.innerHTML = "Turn on";
         document.getElementById("panels").classList.remove("released");
     }
+
 }
 
-function drawPoint(e) {
+function startDraw(e) {
     if (pressed) {
-        context.beginPath();
-        context.arc(e.pageX - myCanvas.offsetLeft, e.pageY - myCanvas.offsetTop, radius, 0, 2 * Math.PI);
-        context.closePath();
-        context.stroke();
-        context.fill();
+        if (currentIndex !== historyChanges) {
+            projectHistory.splice(currentIndex + 1);
+            historyChanges = currentIndex;
+        }
+        drawPoint(radius, e.pageX - myCanvas.offsetLeft, e.pageY - myCanvas.offsetTop);
         context.beginPath();
         context.moveTo(e.pageX - myCanvas.offsetLeft, e.pageY - myCanvas.offsetTop);
-        context.lineWidth = 2 * radius;
         mousePressed = true;
+        isOut = false;
     }
 }
 
 function drawLine(e) {
     if (pressed && mousePressed) {
-        context.lineTo(e.pageX - myCanvas.offsetLeft, e.pageY - myCanvas.offsetTop);
-        context.stroke();
+        if (e.target === myCanvas && isOut === false) {
+            context.lineTo(e.pageX - myCanvas.offsetLeft, e.pageY - myCanvas.offsetTop);
+            context.stroke();
+        }
+        else
+            if (e.target !== myCanvas && isOut === false) {
+                /*let k = (e.pageY - myCanvas.offsetTop - lastY)/(e.pageX - myCanvas.offsetLeft - lastX);
+                let b = lastY - k * lastX;
+                context.lineTo((myCanvas.offsetTop - b)/k, myCanvas.offsetTop + 1);*/
+                context.closePath();
+                isOut = true;
+            }
+            else
+                if (e.target === myCanvas && isOut === true) {
+                    isOut = false;
+                    context.beginPath();
+                    context.moveTo(e.pageX - myCanvas.offsetLeft, e.pageY - myCanvas.offsetTop);
+                }
+        drawPoint(radius, e.pageX - myCanvas.offsetLeft, e.pageY - myCanvas.offsetTop);
+        lastX = e.pageX - myCanvas.offsetLeft;
+        lastY = e.pageY - myCanvas.offsetTop;
     }
 }
 
 function stopDrawing(e) {
-    context.closePath();
-    mousePressed = false;
-    context.lineWidth = 1;
-
+    if (pressed === true && mousePressed === true) {
+        context.closePath();
+        mousePressed = false;
+        isOut = false;
+        var historyImage = new Image();
+        historyImage.src = myCanvas.toDataURL();
+        projectHistory.push(historyImage);
```

Comment:
Can not you use https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getImageData instead of `Image`?